proof/invariant-abstract/ARM_HYP/ArchIpc_AI.thy

```diff
@@ -13,10 +13,36 @@ context Arch begin global_naming ARM_HYP
 named_theorems Ipc_AI_assms
 
 
-crunch pspace_respects_device_region[wp]: set_extra_badge "pspace_respects_device_region"
-
-crunch cap_refs_respects_device_region[wp]: set_extra_badge "cap_refs_respects_device_region"
-  (wp: crunch_wps cap_refs_respects_device_region_dmo)
+lemma update_cap_data_closedform:
+  "update_cap_data pres w cap =
+   (case cap of
+     cap.EndpointCap r badge rights \<Rightarrow>
+       if badge = 0 \<and> \<not> pres then (cap.EndpointCap r (w && mask 28) rights) else cap.NullCap
+   | cap.NotificationCap r badge rights \<Rightarrow>
+       if badge = 0 \<and> \<not> pres then (cap.NotificationCap r (w && mask 28) rights) else cap.NullCap
+   | cap.CNodeCap r bits guard \<Rightarrow>
+       if word_bits < unat ((w >> 3) && mask 5) + bits
+       then cap.NullCap
+       else cap.CNodeCap r bits ((\<lambda>g''. drop (size g'' - unat ((w >> 3) && mask 5)) (to_bl g'')) ((w >> 8) && mask 18))
+   | cap.ThreadCap r \<Rightarrow> cap.ThreadCap r
+   | cap.DomainCap \<Rightarrow> cap.DomainCap
+   | cap.UntypedCap d p n idx \<Rightarrow> cap.UntypedCap d p n idx
+   | cap.NullCap \<Rightarrow> cap.NullCap
+   | cap.ReplyCap t m rights \<Rightarrow> cap.ReplyCap t m rights
+   | cap.IRQControlCap \<Rightarrow> cap.IRQControlCap
+   | cap.IRQHandlerCap irq \<Rightarrow> cap.IRQHandlerCap irq
+   | cap.Zombie r b n \<Rightarrow> cap.Zombie r b n
+   | cap.ArchObjectCap cap \<Rightarrow> cap.ArchObjectCap cap)"
+  apply (cases cap,
+         simp_all only: cap.simps update_cap_data_def is_ep_cap.simps if_False if_True
+                        is_ntfn_cap.simps is_cnode_cap.simps is_arch_cap_def word_size
+                        cap_ep_badge.simps badge_update_def o_def cap_rights_update_def
+                        simp_thms cap_rights.simps Let_def split_def
+                        the_cnode_cap_def fst_conv snd_conv fun_app_def the_arch_cap_def
+                        arch_update_cap_data_def
+                  cong: if_cong)
+  apply (auto simp: word_bits_def)
+  done
```

Comment:
I'm not sure what's going on here. Is the copy in Ipc_A still needed?

Reply:
This patch introduces copies in other arches as well.

Reply:
Well spotted, I’m not sure actually. This might have been an experiment on ARM which then got copied.  Will double check.